compiler/formatter/src/text_edits.rs

```diff
@@ -0,0 +1,126 @@
+use candy_frontend::position::Offset;
+use std::{borrow::Cow, ops::Range};
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct TextEdit {
+    pub range: Range<Offset>,
+    pub new_text: String,
+}
+impl TextEdit {
+    fn is_insert(&self) -> bool {
+        self.range.is_empty()
+    }
+}
+
+/// All text edits ranges refer to positions in the document they are computed on. They therefore
+/// move a document from state S1 to S2 without describing any intermediate state. Text edits ranges
+/// must never overlap, that means no part of the original document must be manipulated by more than
+/// one edit. However, it is possible that multiple edits have the same start position: multiple
+/// inserts, or any number of inserts followed by a single remove or replace edit. If multiple
+/// inserts have the same position, the order in the array defines the order in which the inserted
+/// strings appear in the resulting text.
+///
+/// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textEditArray
+pub struct TextEdits {
+    source: String,
+
+    /// The edits are sorted by their start position.
+    edits: Vec<TextEdit>,
+}
+impl TextEdits {
+    pub fn new(source: String) -> Self {
+        Self {
+            source,
+            edits: vec![],
+        }
+    }
+
+    pub fn source(&self) -> &str {
+        &self.source
+    }
+    pub fn has_edit_at(&self, offset: Offset) -> bool {
+        self.edits
+            .binary_search_by_key(&offset, |it| it.range.start)
+            .map(|_| true) // An edit starts at this position.
+            .unwrap_or_else(|index| {
+                self.edits
+                    .get(index)
+                    // An edit contains this position.
+                    .map(|it| it.range.contains(&offset))
+                    .unwrap_or_default()
+            })
```

Comment:
There's never a run of entries with the same start offset, since they get merged: `change(…)`  checks if the new edit overlaps or touches any existing edit, in which case the code either panics (for overlapping, incompatible changes) or merges them into a single edit